### PR TITLE
[SPARK-28747][SQL] merge the two data source v2 fallback configs

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently we have 2 configs to specify which v2 sources should fallback to v1 code path. One config for read path, and one config for write path.

However, I found it's awkward to work with these 2 configs:
1. for `CREATE TABLE USING format`, should this be read path or write path?
2. for `V2SessionCatalog.loadTable`,  we need to return `UnresolvedTable` if it's a DS v1 or we need to fallback to v1 code path. However, at that time, we don't know if the returned table will be used for read or write.

We don't have any new features or perf improvement in file source v2. The fallback API is just a safeguard if we have bugs in v2 implementations. There are not many benefits to support falling back to v1 for read and write path separately.

This PR proposes to merge these 2 configs into one.

## How was this patch tested?

existing tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -251,64 +251,62 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+    lookupV2Provider() match {
+      // TODO(SPARK-26778): use V2 implementations when partition columns are specified
+      case Some(provider) if partitioningColumns.isEmpty =>
+        saveToV2Source(provider)
+
+      case _ => saveToV1Source()
+    }
+  }
+
+  private def saveToV2Source(provider: TableProvider): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: The code in this method is unchanged.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -835,14 +832,12 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-  private def canUseV2Source(session: SparkSession, providerClass: Class[_]): Boolean = {
-    val useV1Sources =
-      session.sessionState.conf.useV1SourceWriterList.toLowerCase(Locale.ROOT).split(",")
-    val shouldUseV1Source = providerClass.newInstance() match {
-      case d: DataSourceRegister if useV1Sources.contains(d.shortName()) => true
-      case _ => useV1Sources.contains(providerClass.getCanonicalName.toLowerCase(Locale.ROOT))
+  private def lookupV2Provider() = {
+    DataSource.lookupDataSourceV2(source, df.sparkSession.sessionState.conf) match {
+      // File source v2 write path is currently broken.
+      case Some(_: FileDataSourceV2) => None
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This is better than before, which use configs to disable file source v2 write: https://github.com/apache/spark/pull/25465/files#diff-9a6b543db706f1a90f790783d6930a13L1605

configs can be changed, and we shouldn't allow users to use the broken file source v2 write.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: You can review it with while space change ignored if you want to double-check
image

**Comment by cloud-fan** [comment]
cc @brkyvz @rdblue @gengliangwang

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 3]
```diff
@@ -182,21 +180,13 @@ case class DataSourceResolution(
 
   }
 
-  object V1WriteProvider {
-    private val v1WriteOverrideSet =
-      conf.useV1SourceWriterList.toLowerCase(Locale.ROOT).split(",").toSet
-
+  object V1Provider {
     def unapply(provider: String): Option[String] = {
-      if (v1WriteOverrideSet.contains(provider.toLowerCase(Locale.ROOT))) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: We have a bug here. The provider can be a short name or a qualified class name, and `useV1SourceWriterList` should support both. For example, if `orc` is in `useV1SourceWriterList`, then we should fallback to DS V1 if the provider is `orc` or `org.apache.spark.sql.execution.datasources.v2.orc.OrcDataSourceV2`.

This is also how we handle the v1 fallback list in `DataFrameReader/Writer`. See https://github.com/apache/spark/pull/25465/files#diff-f70bda59304588cc3abfa3a9840653f4L211

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/FileDataSourceV2FallBackSuite.scala [hunk 4]
```diff
@@ -130,55 +130,9 @@ class FileDataSourceV2FallBackSuite extends QueryTest with SharedSparkSession {
     }
   }
 
-  test("Fall back to v1 when reading file with write only FileDataSourceV2") {
-    val df = spark.range(10).toDF()
-    withTempPath { file =>
-      val path = file.getCanonicalPath
-      // Dummy File writer should fail as expected.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/FileDataSourceV2FallBackSuite.scala]: Now we always forbid file source v2 in `df.write.save`, this test becomes invalid.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/FileDataSourceV2FallBackSuite.scala [hunk 5]
```diff
@@ -130,55 +130,9 @@ class FileDataSourceV2FallBackSuite extends QueryTest with SharedSparkSession {
...
-          // Writes should fall back to v1 and succeed.
-          df.write.format(dummyParquetWriterV2).save(path)
-          checkAnswer(spark.read.parquet(path), df)
-        }
-      }
-    }
-    withSQLConf(SQLConf.USE_V1_SOURCE_WRITER_LIST.key -> "foo,bar") {
-      withTempPath { file =>
-        val path = file.getCanonicalPath
-        // Dummy File reader should fail as USE_V1_SOURCE_READER_LIST doesn't include it.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/FileDataSourceV2FallBackSuite.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: As a result, I need to update some tests in `PlanResolutionSuite` and `DataSourceV2SQLSuite`, which use qualified class name of orc v2 to trigger v2 code path.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 6]
```diff
@@ -835,14 +832,12 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
 
-  private def canUseV2Source(session: SparkSession, providerClass: Class[_]): Boolean = {
-    val useV1Sources =
-      session.sessionState.conf.useV1SourceWriterList.toLowerCase(Locale.ROOT).split(",")
-    val shouldUseV1Source = providerClass.newInstance() match {
-      case d: DataSourceRegister if useV1Sources.contains(d.shortName()) => true
-      case _ => useV1Sources.contains(providerClass.getCanonicalName.toLowerCase(Locale.ROOT))
+  private def lookupV2Provider() = {
+    DataSource.lookupDataSourceV2(source, df.sparkSession.sessionState.conf) match {
+      // File source v2 write path is currently broken.
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Nit: Add a TODO here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -711,6 +712,23 @@ object DataSource extends Logging {
...
+  /**
+   * Return a [[TableProvider]] instance for the given provider. It returns None if the provider
+   * is not Data Source V2, or it's configured to fallback to Data Source V1 code path.
+   */
+  def lookupDataSourceV2(provider: String, conf: SQLConf): Option[TableProvider] = {
+    val useV1Sources = conf.getConf(SQLConf.USE_V1_SOURCE_LIST).toLowerCase(Locale.ROOT)
+      .split(",").map(_.trim)
+    val cls = lookupDataSource(provider, conf)
+    cls.newInstance() match {
+      case d: DataSourceRegister if useV1Sources.contains(d.shortName()) => None
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: I know this is following the previous logic.
But what if the configuration `USE_V1_SOURCE_LIST` contains the canonical name of the `DataSourceRegister` class.  Maybe we can check this as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 8]
```diff
@@ -711,6 +712,23 @@ object DataSource extends Logging {
     }
   }
 
+  /**
+   * Return a [[TableProvider]] instance for the given provider. It returns None if the provider
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: `Return a [[TableProvider]] instance for the given provider. It returns None if the provider is not Data Source V2, or it's configured to fallback to Data Source V1 code path.`
=>
`Returns an optional [[TableProvider]] instance for the given provider. It returns None if there is no corresponding Data Source V2 implementation, or the provider is configured to fallback to Data Source V1 code path.`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 9]
```diff
@@ -182,21 +180,13 @@ case class DataSourceResolution(
...
-      } else {
-        lazy val providerClass = DataSource.lookupDataSource(provider, conf)
-        provider match {
-          case _ if classOf[TableProvider].isAssignableFrom(providerClass) =>
-            None
-          case _ =>
-            Some(provider)
-        }
+      DataSource.lookupDataSourceV2(provider, conf) match {
+        // Currently file source v2 can't work with tables.
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: ditto

**Review by gengliangwang — COMMENTED** [review]
+1 with the proposal. It makes the fallback simpler.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: It's checked in the next line.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 10]
```diff
@@ -251,64 +251,62 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-      saveToV1Source()
+          case other =>
+            throw new AnalysisException(s"TableProvider implementation $source cannot be " +
+              s"written with $other mode, please use Append or Overwrite " +
+              "modes instead.")
+        }
+
+      // Streaming also uses the data source V2 API. So it may be that the data source
+      // implements v2, but has no v2 implementation for batch writes. In that case, we fall
+      // back to saving as though it's a V1 source.
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: It doesn't look like this is a required change. It just changed the comment indentation and rewrapped. Can you remove this change to avoid future git conflicts?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 11]
```diff
@@ -251,64 +251,62 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+    val sessionOptions = DataSourceV2Utils.extractSessionConfigs(
+      provider, df.sparkSession.sessionState.conf)
+    val options = sessionOptions ++ extraOptions
+    val dsOptions = new CaseInsensitiveStringMap(options.asJava)
+
+    import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
+    provider.getTable(dsOptions) match {
+      // TODO (SPARK-27815): To not break existing tests, here we treat file source as a special
+      // case, and pass the save mode to file source directly. This hack should be removed.
+      case table: FileTable =>
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Is this case still needed, now that `lookupV2Provider` will not return `FileDataSourceV2`?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 12]
```diff
@@ -347,9 +346,16 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSparkSession with Before
...
-      spark.sql(s"CREATE TABLE $catalogName.created USING $orc2 AS SELECT id, data FROM source")
       spark.sql(
-        s"CREATE OR REPLACE TABLE $catalogName.replaced USING $orc2 AS SELECT id, data FROM source")
+        s"""
+           |CREATE TABLE $catalogName.created USING $v2Format
+           |AS SELECT id, data FROM source
+         """.stripMargin)
+      spark.sql(
+        s"""
+           |CREATE TABLE $catalogName.replaced USING $v2Format
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: This changed the SQL. The original query was `CREATE OR REPLACE`.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 13]
```diff
@@ -38,8 +37,7 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSparkSession with Before
 
   import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
 
-  private val orc2 = classOf[OrcDataSourceV2].getName
-  private val v2Source = classOf[FakeV2Provider].getName
+  private val v2Format = classOf[FakeV2Provider].getName
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: Why rename `v2Source` to `v2Format`? That introduces several unnecessary changes.

**Review by keypointt — COMMENTED** [review]
I'm seeing a lot of renaming from `USE_V1_SOURCE_READER_LIST` to `USE_V1_SOURCE_LIST` in test classes, and IMHO better to separate it from logic change PRs :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 14]
```diff
@@ -2333,10 +2325,6 @@ class SQLConf extends Serializable with Logging {
   def continuousStreamingExecutorPollIntervalMs: Long =
     getConf(CONTINUOUS_STREAMING_EXECUTOR_POLL_INTERVAL_MS)
 
-  def useV1SourceReaderList: String = getConf(USE_V1_SOURCE_READER_LIST)
-
-  def useV1SourceWriterList: String = getConf(USE_V1_SOURCE_WRITER_LIST)
-
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: We don't have a policy and it depends on personal preference. I like less code and using the config directly.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 11 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: good point!

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: good catch! must be a mistake when I copy-pasting code.

**Comment by cloud-fan** [comment]
> I'm seeing a lot of renaming from USE_V1_SOURCE_READER_LIST to USE_V1_SOURCE_LIST in test classes

This is the proposed change: merge the confs and have a new name.

**Comment by keypointt** [comment]
> > I'm seeing a lot of renaming from USE_V1_SOURCE_READER_LIST to USE_V1_SOURCE_LIST in test classes
> 
> This is the proposed change: merge the confs and have a new name.

I see. sorry missed it... thank you Wenchen 👍

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 15]
```diff
@@ -251,37 +251,17 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-    val canUseV2 = canUseV2Source(session, cls) && partitioningColumns.isEmpty
-
-    // In Data Source V2 project, partitioning is still under development.
-    // Here we fallback to V1 if partitioning columns are specified.
-    // TODO(SPARK-26778): use V2 implementations when partitioning feature is supported.
-    if (canUseV2) {
-      val provider = cls.getConstructor().newInstance().asInstanceOf[TableProvider]
+    val maybeV2Provider = lookupV2Provider()
+    // TODO(SPARK-26778): use V2 implementations when partition columns are specified
+    if (maybeV2Provider.isDefined && partitioningColumns.isEmpty) {
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This can throw an exception. If the provider is v2 and there is no catalog, then Spark can only append or overwrite (see the cases below). Append and overwrite rely on existing tables and must fail in v2 if the table does not exist. If the user specified partition columns for a table that exists, then it is an error.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Why not add this to `lookupDataSourceV2` instead? Doesn't this logic apply to SQL as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 16]
```diff
@@ -182,21 +180,13 @@ case class DataSourceResolution(
...
-        lazy val providerClass = DataSource.lookupDataSource(provider, conf)
-        provider match {
-          case _ if classOf[TableProvider].isAssignableFrom(providerClass) =>
-            None
-          case _ =>
-            Some(provider)
-        }
+      DataSource.lookupDataSourceV2(provider, conf) match {
+        // TODO(SPARK-28396): Currently file source v2 can't work with tables.
+        case Some(_: FileDataSourceV2) => Some(provider)
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: If this logic were in `lookupDataSourceV2`, I don't think this case would be required.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/FileDataSourceV2FallBackSuite.scala (hunk 4 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/FileDataSourceV2FallBackSuite.scala]: Shouldn't this test always fall back to v1 as a result of the change? And isn't that fallback what this was testing? I think this test should pass and should still be valid.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala [hunk 17]
```diff
@@ -1195,7 +1194,7 @@ class ParquetV2PartitionDiscoverySuite extends ParquetPartitionDiscoverySuite {
   override protected def sparkConf: SparkConf =
     super
       .sparkConf
-      .set(SQLConf.USE_V1_SOURCE_READER_LIST, "")
+      .set(SQLConf.USE_V1_SOURCE_LIST, "")
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala]: What are these tests doing now that `spark.read.parquet` will always use v1?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This doesn't apply to `DataFrameReader`. We can still read file source v2 with `DataFrameReader` and I don't want to break it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 16 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: see https://github.com/apache/spark/pull/25465/files#r316472026

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala (hunk 17 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala]: `spark.read.parquet` can use v1 or v2, `df.write.parquet` always use v1.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 15 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: makes sense. Since it's minor let me just finish this TODO in this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: If the file sources don't support the schema from the metastore, then the read path should not be used. While it doesn't fail while inferring a schema, the source will read using the inferred schema instead of the metastore schema and that's a bug.

For example, if I create a table with 3 columns and drop one, the data and the inferred schema will have the original 3 columns. But the metastore table will have just 2 columns. Reading with v2 would bring back a column that is deleted.

Until the file sources are fixed, they should be disabled in all cases.

**Comment by rdblue** [comment]
@cloud-fan, I commented above, but just to be clear I think that the file source v2 read path also needs to be disabled if it does not use the schema and other settings from the metastore.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I'm talking about `spark.read.parquet(path)`. It's nothing about the metastore, and the behavior is well defined in Spark 2.x:
1. if no schema is specified, e.g. `spark.read.parquet(path)`, then the schema will be inferred.
2. if schema is specified by users, e.g. `spark.read.schema(...).parquet(path)`, then schema inference will be skipped and Spark fails at runtime if the specified schema doesn't match the actual file schema.

I think what you were talking about `CREATE TABLE ... USING parquet`. This is a different case and currently file source v2 is already disabled for it.

**Comment by cloud-fan** [comment]
@rdblue thanks for your review! I replied to your comment, I think we have some misunderstanding about what is the file source read path.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I think you're right that this is not affected by the metastore problem because this is reading a path-based table and not a metastore table.

**Comment by rdblue** [comment]
@cloud-fan, I think you're right. That's not a blocker for this commit.

**Comment by rdblue** [comment]
+1 after rebased and passing tests, assuming no major changes are required.

**Comment by cloud-fan** [comment]
just an import statement conflict in `DataSourceResolution.scala`, easy to fix.

**Review by brkyvz — APPROVED** [review]
LGTM with minor nits

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!
